### DISCUSSION TITLE
CMP-like priors: a prototype

### DISCUSSION THREAD

**Discussion by wlandau** [openpharma/brms.mmrm#92] [Ideas]
This post is a prototype of #40. Borrowing from the methodology of conditional means priors with an identity link function, we put a prior on time differences and get samples of cell means.

## Goals

1.  Define a model formula for a cell means MMRM.
2.  Define a prior on the differences in response from visit to visit.
3.  Fit the model with the prior in (2).
4.  Get posterior samples of the parameters in (1).

## Data

``` r
suppressPackageStartupMessages({
  library(brms)
  library(coda)
  library(emmeans)
  library(mmrm)
  library(posterior)
  library(tidyverse)
  library(zoo)
})
```

We begin with the `FEV` dataset from the `mmrm` package and informally
impute missing values (LOCF then NOCB).

``` r
data(fev_data, package = "mmrm")
data <- fev_data %>%
  mutate(FEV1_CHG = FEV1 - FEV1_BL) %>%
  select(-FEV1) %>%
  group_by(USUBJID) %>%
  complete(
    AVISIT,
    fill = as.list(.[1L, c("ARMCD", "FEV1_BL", "RACE", "SEX", "WEIGHT")])
  ) %>%
  ungroup() %>%
  arrange(USUBJID, AVISIT) %>%
  mutate(USUBJID = as.character(USUBJID)) %>%
  group_by(USUBJID) %>%
  mutate(FEV1_CHG = na.locf(FEV1_CHG, na.rm = FALSE)) %>%
  mutate(FEV1_CHG = na.locf(FEV1_CHG, na.rm = FALSE, fromLast = TRUE)) %>%
  ungroup() %>%
  filter(!is.na(FEV1_CHG))
```

Throughout, we highlight patients 1 (treatment) and 2 (placebo). For
each patient, the 4 study visits are shown in chronological order.

``` r
data %>%
  head(n = 8) %>%
  select(USUBJID, ARMCD, AVISIT, FEV1_CHG)
```

    # A tibble: 8 × 4
      USUBJID ARMCD AVISIT FEV1_CHG
      <chr>   <fct> <fct>     <dbl>
    1 PT1     TRT   VIS1      14.7 
    2 PT1     TRT   VIS2      14.7 
    3 PT1     TRT   VIS3      14.7 
    4 PT1     TRT   VIS4      -4.79
    5 PT2     PBO   VIS1     -13.6 
    6 PT2     PBO   VIS2     -13.6 
    7 PT2     PBO   VIS3      -8.15
    8 PT2     PBO   VIS4       3.78

## Cell means model

We define a simple cell means fixed effect structure. This is the
parameterization where we want to do inference.

``` r
formula_cell_means <- brmsformula(FEV1_CHG ~ 0 + AVISIT:ARMCD)
```

``` r
formula_cell_means
```

    FEV1_CHG ~ 0 + AVISIT:ARMCD 

We obtain the `brms` model matrix for the cell means parameterization.

``` r
model_matrix_cell_means <- make_standata(
  formula = formula_cell_means,
  data = data
)$X
```

``` r
colnames(model_matrix_cell_means)
```

    [1] "AVISITVIS1:ARMCDPBO" "AVISITVIS2:ARMCDPBO" "AVISITVIS3:ARMCDPBO"
    [4] "AVISITVIS4:ARMCDPBO" "AVISITVIS1:ARMCDTRT" "AVISITVIS2:ARMCDTRT"
    [7] "AVISITVIS3:ARMCDTRT" "AVISITVIS4:ARMCDTRT"

Below, the first four rows are for patient 1 (treatment), and the next
four rows are for patient 2 (placebo).

``` r
head(unname(model_matrix_cell_means), n = 8)
```

         [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
    [1,]    0    0    0    0    1    0    0    0
    [2,]    0    0    0    0    0    1    0    0
    [3,]    0    0    0    0    0    0    1    0
    [4,]    0    0    0    0    0    0    0    1
    [5,]    1    0    0    0    0    0    0    0
    [6,]    0    1    0    0    0    0    0    0
    [7,]    0    0    1    0    0    0    0    0
    [8,]    0    0    0    1    0    0    0    0

## Reparameterize to differences

Now we create an 8x8 transformation matrix to map the cell means model
matrix to a time difference model matrix. We will put an informative
prior on the time difference parameterization.

``` r
cell_means <- colnames(model_matrix_cell_means)
transform <- NULL
for (cell_mean in cell_means) {
  group <- cell_mean %>%
    gsub(pattern = "^.*:ARMCD", replacement = "")
  time <- cell_mean %>%
    gsub(pattern = "^AVISITVIS", replacement = "") %>%
    gsub(pattern = ":.*$", replacement = "") %>%
    as.integer()
  zeroes <- rep(0, 4)
  differences <- as.integer(seq_len(4) <= time)
  if (group == "PBO") {
    column <- c(differences, zeroes)
  } else {
    column <- c(zeroes, differences)
  }
  names(column) <- c(
    paste0("PBO_", c("1", "1_to_2", "2_to_3", "3_to_4")),
    paste0("TRT_", c("1", "1_to_2", "2_to_3", "3_to_4"))
  )
  transform <- cbind(transform, column)
}
colnames(transform) <- c(
  paste0("PBO:", seq_len(4)),
  paste0("TRT:", seq_len(4))
)
```

``` r
transform
```

               PBO:1 PBO:2 PBO:3 PBO:4 TRT:1 TRT:2 TRT:3 TRT:4
    PBO_1          1     1     1     1     0     0     0     0
    PBO_1_to_2     0     1     1     1     0     0     0     0
    PBO_2_to_3     0     0     1     1     0     0     0     0
    PBO_3_to_4     0     0     0     1     0     0     0     0
    TRT_1          0     0     0     0     1     1     1     1
    TRT_1_to_2     0     0     0     0     0     1     1     1
    TRT_2_to_3     0     0     0     0     0     0     1     1
    TRT_3_to_4     0     0     0     0     0     0     0     1

The model matrix for time differences is:

``` r
model_matrix_differences <- model_matrix_cell_means %*% t(transform)
rownames(model_matrix_differences) <- paste(
  data$USUBJID,
  data$ARMCD,
  gsub("^VIS", "", data$AVISIT),
  sep = "|"
)
```

Within each treatment group, each visit is modeled as a cumulative sum
of consecutive time differences. The row names show factors from the
data, and the column names are model terms.

``` r
head(model_matrix_differences, 8)
```

              PBO_1 PBO_1_to_2 PBO_2_to_3 PBO_3_to_4 TRT_1 TRT_1_to_2 TRT_2_to_3 TRT_3_to_4
    PT1|TRT|1     0          0          0          0     1          0          0          0
    PT1|TRT|2     0          0          0          0     1          1          0          0
    PT1|TRT|3     0          0          0          0     1          1          1          0
    PT1|TRT|4     0          0          0          0     1          1          1          1
    PT2|PBO|1     1          0          0          0     0          0          0          0
    PT2|PBO|2     1          1          0          0     0          0          0          0
    PT2|PBO|3     1          1          1          0     0          0          0          0
    PT2|PBO|4     1          1          1          1     0          0          0          0

We can recover the original model matrix by inverting the
transformation. This will be useful later when we derive posterior
samples of cell means from posterior samples of time differences.

``` r
head(model_matrix_differences %*% t(solve(transform)), 8)
```

              PBO:1 PBO:2 PBO:3 PBO:4 TRT:1 TRT:2 TRT:3 TRT:4
    PT1|TRT|1     0     0     0     0     1     0     0     0
    PT1|TRT|2     0     0     0     0     0     1     0     0
    PT1|TRT|3     0     0     0     0     0     0     1     0
    PT1|TRT|4     0     0     0     0     0     0     0     1
    PT2|PBO|1     1     0     0     0     0     0     0     0
    PT2|PBO|2     0     1     0     0     0     0     0     0
    PT2|PBO|3     0     0     1     0     0     0     0     0
    PT2|PBO|4     0     0     0     1     0     0     0     0

## Model the differences

We define an alternative dataset with the new parameterization and
subject visit indicators.

``` r
data_differences <- model_matrix_differences %>%
  as.data.frame() %>%
  as_tibble() %>%
  mutate(
    FEV1_CHG = data$FEV1_CHG,
    USUBJID = data$USUBJID,
    AVISIT = data$AVISIT
  )
```

We define an MMRM model formula using the difference terms as fixed
effects and subject visit indicators in the correlation structure.

``` r
formula_text <- paste(
  "FEV1_CHG ~ 0 +",
  paste(colnames(model_matrix_differences), collapse = " + "),
  "+ ar(time = AVISIT, gr = USUBJID)"
)
formula_differences <- brmsformula(as.formula(formula_text))
```

``` r
formula_differences
```

    FEV1_CHG ~ 0 + PBO_1 + PBO_1_to_2 + PBO_2_to_3 + PBO_3_to_4 + TRT_1 + TRT_1_to_2 + TRT_2_to_3 + TRT_3_to_4 + ar(time = AVISIT, gr = USUBJID) 

We manually set a prior on each of the treatment differences. In
practice, this will be informed by historical data on the specific
treatment differences and not simply standard normal.

``` r
prior_differences <- do.call(
  c,
  map(
    colnames(model_matrix_differences),
    ~prior_string(prior = "normal(0, 1)", class = "b", coef = .x)
  )
)
```

``` r
prior_differences
```

            prior class       coef group resp dpar nlpar   lb   ub source
     normal(0, 1)     b      PBO_1                       <NA> <NA>   user
     normal(0, 1)     b PBO_1_to_2                       <NA> <NA>   user
     normal(0, 1)     b PBO_2_to_3                       <NA> <NA>   user
     normal(0, 1)     b PBO_3_to_4                       <NA> <NA>   user
     normal(0, 1)     b      TRT_1                       <NA> <NA>   user
     normal(0, 1)     b TRT_1_to_2                       <NA> <NA>   user
     normal(0, 1)     b TRT_2_to_3                       <NA> <NA>   user
     normal(0, 1)     b TRT_3_to_4                       <NA> <NA>   user

We fit an MMRM using this special formula, data, and prior.

``` r
model_differences <- brm(
  formula = formula_differences,
  data = data_differences,
  prior = prior_differences,
  refresh = 0
)
```

## Samples of cell means

Extract the posterior draws of the differences from the model.

``` r
draws_differences <- model_differences %>%
  as_tibble() %>%
  select(starts_with("b_"), -starts_with("b_sigma"))
```

``` r
draws_differences
```

    # A tibble: 4,000 × 8
       b_PBO_1 b_PBO_1_to_2 b_PBO_2_to_3 b_PBO_3_to_4 b_TRT_1 b_TRT_1_to_2
         <dbl>        <dbl>        <dbl>        <dbl>   <dbl>        <dbl>
     1   -3.13        1.85          1.81         2.87 -0.149          2.04
     2   -3.63        1.43          2.09         2.74 -0.309          1.64
     3   -4.02        2.11          2.16         2.78 -1.22           1.67
     4   -2.54        0.532         2.53         3.60  0.216          1.83
     5   -3.90        1.77          1.72         1.97  0.329          1.22
     6   -2.25        0.898         2.23         3.85 -0.307          1.81
     7   -3.99        1.55          1.70         1.82  0.543          1.38
     8   -3.20        1.06          1.81         2.07  0.577          1.45
     9   -2.66        0.737         2.12         3.14  0.0580         1.26
    10   -2.22        2.29          1.42         2.96  0.700          1.37
    # ℹ 3,990 more rows
    # ℹ 2 more variables: b_TRT_2_to_3 <dbl>, b_TRT_3_to_4 <dbl>

We invert the transformation to derive posterior samples of cell means.

``` r
draws_cell_means <- as.matrix(draws_differences) %*% t(solve(transform)) %>%
  as_tibble()
```

``` r
draws_cell_means
```

    # A tibble: 4,000 × 8
       `PBO:1` `PBO:2` `PBO:3` `PBO:4` `TRT:1` `TRT:2` `TRT:3` `TRT:4`
         <dbl>   <dbl>   <dbl>   <dbl>   <dbl>   <dbl>   <dbl>   <dbl>
     1   -4.98  0.0439  -1.06     2.87  -2.18   0.285  -2.37      4.12
     2   -5.06 -0.662   -0.644    2.74  -1.95   0.191  -2.65      4.10
     3   -6.14 -0.0434  -0.625    2.78  -2.89   0.588  -2.77      3.85
     4   -3.07 -2.00    -1.07     3.60  -1.62  -0.500  -2.98      5.31
     5   -5.68  0.0579  -0.252    1.97  -0.889 -1.08   -0.144     2.45
     6   -3.15 -1.33    -1.62     3.85  -2.12   0.0540 -3.35      5.11
     7   -5.54 -0.151   -0.123    1.82  -0.840 -1.01   -0.0772    2.47
     8   -4.26 -0.748   -0.261    2.07  -0.875 -0.825  -1.03      3.31
     9   -3.39 -1.38    -1.02     3.14  -1.21   0.337  -3.64      4.56
    10   -4.51  0.865   -1.53     2.96  -0.672 -0.244  -2.26      3.88
    # ℹ 3,990 more rows

If we had covariates and/or a non-cell-means parameterization to begin with, we could apply the custom marginal means technique `brms.mmrm` already implements for #53.

## Challenges

It is easy to work with the cell means parameterization, but what if the user begins with a different base model? And with baseline covariates and interactions with baseline, how do we avoid implicitly conditioning on the a reference level at a subset of the data? Figuring out these questions is the next step of this prototype. I hope a combination of workarounds, guardrails, and compromises can give us something powerful and useful for `brms.mmrm`.

**Comment by chstock** [comment]
This is great! Looking forward to discussing the next steps!

**Comment by yonicd** [comment]
This looks great!

in the past i have had to construct mean over time points as the endpoint, eg if t timepoints then each gets 1/t weight instead of 1 (or even a nonlin functional if the clinical team is feeling like they want to test a hypothesis) would this be possible in this construction?

**Reply by wlandau** [reply]
Conditional means priors work best when the transformation matrix is full rank, and averaging across time points might be tricky. I think we would have to set one parameter to be the average and the others to be time-specific contrasts on the average. Kind of like "sum contrasts" from `stats::contr.sum()`.

**Comment by andrew-bean** [comment]
Nice work as usual @wlandau !

Still thinking about @weberse2 's suggestion in #40. IMO setting the contrast attribute of the AVISIT factor is a simpler way to reparametrize the model and specify the prior on some alternative linear function of the model coefficients. 

In the code above we need to specify a new formula which does not resemble the usual MMRM `formula_differences` to define the model in terms of the differences. We also need to manually construct the matrix `transform` to do inference on the cell means. 

By contrast (sorry), if you simply set the contrast attribute for AVISIT, you avoid both these inconveniences. With the alternative contrast attribute on AVISIT, brms will treat this as a reparametrized model, even though the formula is unchanged from the typical MMRM specification. Basic R functions like lm behave in a similar way. In brms, the consequence is you scan specify your prior on the forward difference contrasts rather than the cell means, without writing a completely new model.

The syntax would be something like this:

```r


suppressPackageStartupMessages({
  library(brms)
  library(coda)
  library(emmeans)
  library(mmrm)
  library(posterior)
  library(tidyverse)
  library(zoo)
})

data(fev_data, package = "mmrm")
data <- fev_data %>%
  mutate(FEV1_CHG = FEV1 - FEV1_BL) %>%
  select(-FEV1) %>%
  group_by(USUBJID) %>%
  complete(
    AVISIT,
    fill = as.list(.[1L, c("ARMCD", "FEV1_BL", "RACE", "SEX", "WEIGHT")])
  ) %>%
  ungroup() %>%
  arrange(USUBJID, AVISIT) %>%
  mutate(USUBJID = as.character(USUBJID)) %>%
  group_by(USUBJID) %>%
  mutate(FEV1_CHG = na.locf(FEV1_CHG, na.rm = FALSE)) %>%
  mutate(FEV1_CHG = na.locf(FEV1_CHG, na.rm = FALSE, fromLast = TRUE)) %>%
  ungroup() %>%
  filter(!is.na(FEV1_CHG))


formula_cell_means <- brmsformula(FEV1_CHG ~ 0 + AVISIT : ARMCD)

# we can just use a typical MMRM formula here 
formula_sdif <- brmsformula(FEV1_CHG ~ AVISIT * ARMCD)

# form forward-difference contrasts
cc <- MASS::contr.sdif(nlevels(data$AVISIT))
cc

# must say I have not wrapped my head around the output from this
# MASS function. Its Moore-Penrose G-inverse makes some sense, in that it 
# maps visits to visit differences.
zapsmall(MASS::ginv(cc))

# but this seems to work:
cc_alt <- diag(1, nlevels(data$AVISIT))
colnames(cc_alt) <- c("1", colnames(cc))
rownames(cc_alt) <- levels(data$AVISIT)
cc_alt[lower.tri(cc_alt)] <- 1
cc_alt
solve(cc_alt)

# new data frame in which nothing has changed except the contrasts attribute
# of the AVISIT factor
data_sdif <- data
contrasts(data_sdif$AVISIT) <- cc_alt[,-1] # must drop the intercept from the contrast definition

model_matrix_cell_means <- make_standata(
  formula = formula_cell_means,
  data = data
)$X
model_matrix_sdif <- make_standata(
  formula = formula_sdif,
  data = data_sdif
)$X

head(model_matrix_cell_means, n = 8)
head(model_matrix_sdif, n = 8)

get_prior(formula_cell_means, data = data)
# prior now specified for the reparametrized coefficients, even though the
# formula is unchanged
get_prior(formula_sdif, data = data_sdif)

# # cell means model
# fit_cell_means <- brm(formula_cell_means, data, sample_prior = "only",
#                       prior = c(prior(normal(0, 1), class = "b")))

# reparametrized model in terms of successive differences in visits
fit_sdif <- brm(formula_sdif, data_sdif, sample_prior = "only",
                prior = c(prior(normal(0, 1), class = "b")))

# prior draws for the coefficients under each prior specification
# beta_cell_means <- as.matrix(fit_cell_means) 
beta_sdif <- as.matrix(fit_sdif)

# beta_cell_means <- beta_cell_means[, grep("^b_*", colnames(beta_cell_means))]
beta_sdif <- beta_sdif[, grep("^b_*", colnames(beta_sdif))]

# to do inference on the cell means, you could just do something like this
# (no need for a transformation)

# define the cells
cells <- data_sdif %>%
  select(ARMCD, AVISIT) %>%
  arrange(ARMCD, AVISIT) %>%
  distinct() %>%
  mutate(FEV1_CHG = 0,
         cell_label = interaction(ARMCD, AVISIT))

# posterior draws for the linear predictor at these levels
draws <- posterior_linpred(fit_sdif, newdata = cells)

```

**Reply by wlandau** [reply]
Thanks for posting, @andrew-bean. As we discussed today, a contrast-based implementation looks far cleaner and more achievable than manually constructing the model matrix from scratch. I think we may still need to use model matrices (from `brms::make_standata()`) for validation and post-hoc parameter transformations, but this is much easier than manually building those model matrices to begin with.

**Reply by wlandau** [reply]
Good news: the G-inverse should let us easily transform posterior samples from sdif to cell means:

```r
transform_sdif_to_cells <- MASS::ginv(model_matrix_sdif) %*%
  model_matrix_cell_means
transformed <- model_matrix_sdif %*% transform_sdif_to_cells
max(abs(model_matrix_cell_means - transformed))
#> [1] 1.239092e-14
```

A stumbling block: like you mentioned in our meeting yesterday @andrew-bean, the sdif parameterization is treatment-effect-like and not cell-means-like. In other words, for the treatment group, each successive difference is defined relative to the placebo group. As you said, this may make it difficult to set priors in an intuitive way.

```r
> head(model_matrix_sdif, 8)
  (Intercept) AVISIT1 AVISIT2 AVISIT3 ARMCDTRT AVISIT1:ARMCDTRT AVISIT2:ARMCDTRT AVISIT3:ARMCDTRT
1           1       0       0       0        1                0                0                0
2           1       1       0       0        1                1                0                0
3           1       1       1       0        1                1                1                0
4           1       1       1       1        1                1                1                1
5           1       0       0       0        0                0                0                0
6           1       1       0       0        0                0                0                0
7           1       1       1       0        0                0                0                0
8           1       1       1       1        0                0                0                0
```

I have experimented with many different contrast matrices for `ARMCD`, and I can't seem to find one that produces this, which is what we want:

```r
  (Intercept) AVISIT1 AVISIT2 AVISIT3 ARMCDTRT AVISIT1:ARMCDTRT AVISIT2:ARMCDTRT AVISIT3:ARMCDTRT
1           0       0       0       0        1                0                0                0
2           0       0       0       0        1                1                0                0
3           0       0       0       0        1                1                1                0
4           0       0       0       0        1                1                1                1
5           1       0       0       0        0                0                0                0
6           1       1       0       0        0                0                0                0
7           1       1       1       0        0                0                0                0
8           1       1       1       1        0                0                0                0
```

Unfortunately, unless we can find a contrast matrix that produces this model matrix, the behind-the-scenes mechanism might have to manually construct said model matrix and define a special formula to go with it.

**Reply by wlandau** [reply]
> Unfortunately, unless we can find a contrast matrix that produces this model matrix, the behind-the-scenes mechanism might have to manually construct said model matrix and define a special formula to go with it.

As it turns out, the Kronecker product makes this easy:

```r
matrix_group <- model.matrix(~ 0 + ARMCD, filter(data, AVISIT == AVISIT[1]))
matrix_time <- diag(4) + lower.tri(diag(4))
matrix_sdif <- kronecker(matrix_treatment, matrix_time)
colnames(matrix_sdif) <- c(paste0("PBO|", seq_len(4)), paste0("TRT|", seq_len(4)))
head(matrix_sdif, n = 8)
#>      PBO|1 PBO|2 PBO|3 PBO|4 TRT|1 TRT|2 TRT|3 TRT|4
#> [1,]     0     0     0     0     1     0     0     0
#> [2,]     0     0     0     0     1     1     0     0
#> [3,]     0     0     0     0     1     1     1     0
#> [4,]     0     0     0     0     1     1     1     1
#> [5,]     1     0     0     0     0     0     0     0
#> [6,]     1     1     0     0     0     0     0     0
#> [7,]     1     1     1     0     0     0     0     0
#> [8,]     1     1     1     1     0     0     0     0
```

**Reply by wlandau** [reply]
> the sdif parameterization is treatment-effect-like and not cell-means-like. In other words, for the treatment group, each successive difference is defined relative to the placebo group. As you said, this may make it difficult to set priors in an intuitive way.

We might actually want treatment-effect-like sdif in certain situations. It depends on whether we want to model the treatment groups as independent in the prior, or if we want them correlated. This could be weighed against intuition on a case-by-case basis.

**Reply by wlandau** [reply]
> We might actually want treatment-effect-like sdif in certain situations. It depends on whether we want to model the treatment groups as independent in the prior, or if we want them correlated. This could be weighed against intuition on a case-by-case basis.

Implemented in #100 via `brm_archetype_successive_effects()`.

On the original thread, I think we landed well in #96 and #100 with informative prior archetypes.
